# Supervised Classification Baselines on MATH Dataset - Results

This notebook displays the results from fine-tuning experiments on the classification task using the MATH dataset.

In [1]:
import pandas as pd
import os

pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

RESULTS_PATH = "../experiment_results/classification/0_supervised_baselines_results"

try:
    os.chdir(RESULTS_PATH)
    print(os.getcwd())
except Exception:
    raise ValueError(f"Experiment results path not found: {RESULTS_PATH}")

/Users/yacinebenihaddadene/Documents/School/S9/NLP1/Project/FewShotOptimisation/AdvancedNLP-Project/experiment_results/classification/0_supervised_baselines_results


# Unsloth Generative Fine-Tuning

In [2]:
df = pd.read_csv('experiments.csv')

print(f"Total experiments: {len(df)}")
df

Total experiments: 11


timestamp                                model_path  \
0   2026-01-13 18:22:25   fine_tunings/unsloth/finetune_llama3_8b   
1   2026-01-13 19:23:12    fine_tunings/unsloth/finetune_gemma_7b   
2   2026-01-13 19:33:17   fine_tunings/unsloth/finetune_llama3_8b   
3   2026-01-13 19:43:23  fine_tunings/unsloth/finetune_mistral_7b   
4   2026-01-13 19:51:18   fine_tunings/unsloth/finetune_phi3_mini   
5   2026-01-13 19:57:23    fine_tunings/unsloth/finetune_qwen2_7b   
6   2026-01-13 20:12:39                 unsloth/gemma-7b-bnb-4bit   
7   2026-01-13 20:21:49               unsloth/llama-3-8b-bnb-4bit   
8   2026-01-13 20:31:00          unsloth/mistral-7b-v0.3-bnb-4bit   
9   2026-01-13 20:37:25   unsloth/Phi-3-mini-4k-instruct-bnb-4bit   
10  2026-01-13 20:43:17                 unsloth/Qwen2-7B-bnb-4bit   

                                 model_name             run_name  \
0                                 llama3:8b   finetune_llama3_8b   
1                                  gemma:7b    finetune_gemma_7b   
2                                 llama3:8b   finetune_llama3_8b   
3                                mistral:7b  finetune_mistral_7b   
4                                 phi3:mini   finetune_phi3_mini   
5                                  qwen2:7b    finetune_qwen2_7b   
6                 unsloth/gemma-7b-bnb-4bit      base_model_eval   
7               unsloth/llama-3-8b-bnb-4bit      base_model_eval   
8          unsloth/mistral-7b-v0.3-bnb-4bit      base_model_eval   
9   unsloth/Phi-3-mini-4k-instruct-bnb-4bit      base_model_eval   
10                unsloth/Qwen2-7B-bnb-4bit      base_model_eval   

                      dataset  test_size  accuracy  precision_weighted  \
0   qwedsacf/competition_math       1250    0.7760            0.827533   
1   qwedsacf/competition_math       1250    0.0000            0.000000   
2   qwedsacf/competition_math       1250    0.7760            0.827533   
3   qwedsacf/competition_math       1250    0.8136            0.842657   
4   qwedsacf/competition_math       1250    0.6408            0.735949   
5   qwedsacf/competition_math       1250    0.7528            0.795759   
6   qwedsacf/competition_math       1250    0.0000            0.000000   
7   qwedsacf/competition_math       1250    0.3624            0.479597   
8   qwedsacf/competition_math       1250    0.0336            0.114706   
9   qwedsacf/competition_math       1250    0.0000            0.000000   
10  qwedsacf/competition_math       1250    0.4448            0.436731   

    recall_weighted  f1_weighted  tokens_per_sec  num_epochs  batch_size  \
0            0.7760     0.789775       21.270479           3           2   
1            0.0000     0.000000       19.265760           3           2   
2            0.7760     0.789775       21.294834           3           2   
3            0.8136     0.822848       21.344999           3           2   
4            0.6408     0.669414       27.218389           3           2   
5            0.7528     0.759331       15.286895           3           2   
6            0.0000     0.000000       18.833848           0           2   
7            0.3624     0.334838       21.966956           0           2   
8            0.0336     0.048908       22.411656           0           2   
9            0.0000     0.000000       37.977378           0           2   
10           0.4448     0.407763       13.626541           0           2   

    learning_rate  lora_rank  lora_alpha  Algebra_precision  Algebra_recall  \
0          0.0002         16          16           0.762611        0.924460   
1          0.0002         16          16           0.000000        0.000000   
2          0.0002         16          16           0.762611        0.924460   
3          0.0002         16          16           0.912500        0.787770   
4          0.0002         16          16           0.852761        0.500000   
5          0.0002         16          16           0.926966        0.593525   
6          0.

## Summary Metrics

In [3]:
# Key metrics comparison
summary_cols = ['model_path', 'dataset', 'test_size', 'accuracy', 'f1_weighted', 'precision_weighted', 'recall_weighted', 'tokens_per_sec']
available_cols = [c for c in summary_cols if c in df.columns]
df[available_cols].sort_values('accuracy', ascending=False)

model_path                    dataset  \
3   fine_tunings/unsloth/finetune_mistral_7b  qwedsacf/competition_math   
0    fine_tunings/unsloth/finetune_llama3_8b  qwedsacf/competition_math   
2    fine_tunings/unsloth/finetune_llama3_8b  qwedsacf/competition_math   
5     fine_tunings/unsloth/finetune_qwen2_7b  qwedsacf/competition_math   
4    fine_tunings/unsloth/finetune_phi3_mini  qwedsacf/competition_math   
10                 unsloth/Qwen2-7B-bnb-4bit  qwedsacf/competition_math   
7                unsloth/llama-3-8b-bnb-4bit  qwedsacf/competition_math   
8           unsloth/mistral-7b-v0.3-bnb-4bit  qwedsacf/competition_math   
1     fine_tunings/unsloth/finetune_gemma_7b  qwedsacf/competition_math   
6                  unsloth/gemma-7b-bnb-4bit  qwedsacf/competition_math   
9    unsloth/Phi-3-mini-4k-instruct-bnb-4bit  qwedsacf/competition_math   

    test_size  accuracy  f1_weighted  precision_weighted  recall_weighted  \
3        1250    0.8136     0.822848            0.842657           0.8136   
0        1250    0.7760     0.789775            0.827533           0.7760   
2        1250    0.7760     0.789775            0.827533           0.7760   
5        1250    0.7528     0.759331            0.795759           0.7528   
4        1250    0.6408     0.669414            0.735949           0.6408   
10       1250    0.4448     0.407763            0.436731           0.4448   
7        1250    0.3624     0.334838            0.479597           0.3624   
8        1250    0.0336     0.048908            0.114706           0.0336   
1        1250    0.0000     0.000000            0.000000           0.0000   
6        1250    0.0000     0.000000            0.000000           0.0000   
9        1250    0.0000     0.000000            0.000000           0.0000   

    tokens_per_sec  
3        21.344999  
0        21.270479  
2        21.294834  
5        15.286895  
4        27.218389  
10       13.626541  
7        21.966956  
8        22.411656  
1        19.265760  
6        18.833848  
9        37.977378

## Per-Class F1 Scores

In [4]:
# Extract per-class F1 columns
f1_cols = [c for c in df.columns if c.endswith('_f1') and c != 'f1_weighted']
print(f"Found {len(f1_cols)} per-class F1 columns.")
if f1_cols:
    display_df = df[['model_path', 'dataset', 'test_size',] + f1_cols].copy()
    display_df.columns = ['model_path', 'dataset', 'test_size',] + [c.replace('_f1', '') for c in f1_cols]
    display(display_df)

Found 7 per-class F1 columns.


model_path                    dataset  \
0    fine_tunings/unsloth/finetune_llama3_8b  qwedsacf/competition_math   
1     fine_tunings/unsloth/finetune_gemma_7b  qwedsacf/competition_math   
2    fine_tunings/unsloth/finetune_llama3_8b  qwedsacf/competition_math   
3   fine_tunings/unsloth/finetune_mistral_7b  qwedsacf/competition_math   
4    fine_tunings/unsloth/finetune_phi3_mini  qwedsacf/competition_math   
5     fine_tunings/unsloth/finetune_qwen2_7b  qwedsacf/competition_math   
6                  unsloth/gemma-7b-bnb-4bit  qwedsacf/competition_math   
7                unsloth/llama-3-8b-bnb-4bit  qwedsacf/competition_math   
8           unsloth/mistral-7b-v0.3-bnb-4bit  qwedsacf/competition_math   
9    unsloth/Phi-3-mini-4k-instruct-bnb-4bit  qwedsacf/competition_math   
10                 unsloth/Qwen2-7B-bnb-4bit  qwedsacf/competition_math   

    test_size   Algebra  Counting & Probability  Geometry  \
0        1250  0.835772                0.765217  0.689956   
1        1250  0.000000                0.000000  0.000000   
2        1250  0.835772                0.765217  0.689956   
3        1250  0.845560                0.805085  0.741071   
4        1250  0.630385                0.727273  0.595041   
5        1250  0.723684                0.725490  0.694915   
6        1250  0.000000                0.000000  0.000000   
7        1250  0.377483                0.481675  0.440559   
8        1250  0.000000                0.000000  0.028369   
9        1250  0.000000                0.000000  0.000000   
10       1250  0.468917                0.602317  0.612565   

    Intermediate Algebra  Number Theory  Prealgebra  Precalculus  
0               0.882759       0.832827    0.596591     0.933333  
1               0.000000       0.000000    0.000000     0.000000  
2               0.882759       0.832827    0.596591     0.933333  
3               0.929336       0.800000    0.681818     0.953975  
4               0.765432       0.716981    0.554502     0.747573  
5               0.835938       0.800000    0.682540     0.885496  
6               0.000000       0.000000    0.000000     0.000000  
7               0.337568       0.550239    0.164062     0.016393  
8               0.000000       0.000000    0.132530     0.235294  
9               0.000000       0.000000    0.000000     0.000000  
10              0.390135       0.628726    0.194757     0.000000

## Training Configuration

In [5]:
# Config columns
config_cols = ['model_path', 'model_name', 'num_epochs', 'batch_size', 'learning_rate', 'lora_rank', 'lora_alpha']
available_config = [c for c in config_cols if c in df.columns]
df[available_config]

model_path  \
0    fine_tunings/unsloth/finetune_llama3_8b   
1     fine_tunings/unsloth/finetune_gemma_7b   
2    fine_tunings/unsloth/finetune_llama3_8b   
3   fine_tunings/unsloth/finetune_mistral_7b   
4    fine_tunings/unsloth/finetune_phi3_mini   
5     fine_tunings/unsloth/finetune_qwen2_7b   
6                  unsloth/gemma-7b-bnb-4bit   
7                unsloth/llama-3-8b-bnb-4bit   
8           unsloth/mistral-7b-v0.3-bnb-4bit   
9    unsloth/Phi-3-mini-4k-instruct-bnb-4bit   
10                 unsloth/Qwen2-7B-bnb-4bit   

                                 model_name  num_epochs  batch_size  \
0                                 llama3:8b           3           2   
1                                  gemma:7b           3           2   
2                                 llama3:8b           3           2   
3                                mistral:7b           3           2   
4                                 phi3:mini           3           2   
5                                  qwen2:7b           3           2   
6                 unsloth/gemma-7b-bnb-4bit           0           2   
7               unsloth/llama-3-8b-bnb-4bit           0           2   
8          unsloth/mistral-7b-v0.3-bnb-4bit           0           2   
9   unsloth/Phi-3-mini-4k-instruct-bnb-4bit           0           2   
10                unsloth/Qwen2-7B-bnb-4bit           0           2   

    learning_rate  lora_rank  lora_alpha  
0          0.0002         16          16  
1          0.0002         16          16  
2          0.0002         16          16  
3          0.0002         16          16  
4          0.0002         16          16  
5          0.0002         16          16  
6          0.0002         16          16  
7          0.0002         16          16  
8          0.0002         16          16  
9          0.0002         16          16  
10         0.0002         16          16

# All Experiments: Generative + Discriminative + Zero-shot base models Baselines

In [6]:
import yaml

# Load evaluation results
yaml_path = 'evaluation_results.yaml'

try:
    with open(yaml_path, 'r') as f:
        eval_data = yaml.safe_load(f)

    eval_df = pd.DataFrame(eval_data)

    # improved display
    display_cols = ['Model', 'Finetune', 'Accuracy', 'F1_Weighted', 'Avg_Latency_Seconds', 'Task']
    print("Evaluation Results:")
    display(eval_df[display_cols].sort_values('Accuracy', ascending=False))

    # Grouped analysis
    print("\nAverage Performance by Fine-tune Type:")
    display(eval_df.groupby('Finetune')[['Accuracy', 'F1_Weighted', 'Avg_Latency_Seconds']].mean(numeric_only=True))

except FileNotFoundError:
    print(f"Could not find file at {yaml_path}")

Evaluation Results:


Model  \
1                                  qwen2:7b   
2                                 llama3:8b   
5                                mistral:7b   
0                                mistral:7b   
4                                 llama3:8b   
7                                  qwen2:7b   
6                                 phi3:mini   
12                unsloth/Qwen2-7B-bnb-4bit   
9               unsloth/llama-3-8b-bnb-4bit   
10         unsloth/mistral-7b-v0.3-bnb-4bit   
3                                  gemma:7b   
8                 unsloth/gemma-7b-bnb-4bit   
11  unsloth/Phi-3-mini-4k-instruct-bnb-4bit   

                                       Finetune  Accuracy  F1_Weighted  \
1   classification head classification finetune    0.8512     0.849699   
2   classification head classification finetune    0.8328     0.832554   
5            generative classification finetune    0.8136     0.822848   
0   classification head classification finetune    0.8080     0.805375   
4            generative classification finetune    0.7760     0.789775   
7            generative classification finetune    0.7528     0.759331   
6            generative classification finetune    0.6408     0.669414   
12                                         None    0.4448     0.407763   
9                                          None    0.3624     0.334838   
10                                         None    0.0336     0.048908   
3            generative classification finetune    0.0000     0.000000   
8                                          None    0.0000     0.000000   
11                                         None    0.0000     0.000000   

    Avg_Latency_Seconds                                 Task  
1              0.144110             FineTuned classification  
2              0.156486             FineTuned classification  
5              0.468494             FineTuned classification  
0              0.177346             FineTuned classification  
4              0.469597             FineTuned classification  
7              0.292904             FineTuned classification  
6              0.367399             FineTuned classification  
12             0.210061  Zero-shot generative classification  
9              0.426568  Zero-shot generative classification  
10             0.420139  Zero-shot generative classification  
3              0.518889             FineTuned classification  
8              0.447280  Zero-shot generative classification  
11             0.263315  Zero-shot generative classification


Average Performance by Fine-tune Type:


Accuracy  F1_Weighted  \
Finetune                                                             
None                                         0.168160     0.158302   
classification head classification finetune  0.830667     0.829209   
generative classification finetune           0.596640     0.608274   

                                             Avg_Latency_Seconds  
Finetune                                                          
None                                                    0.353472  
classification head classification finetune             0.159314  
generative classification finetune                      0.423457

### Analysis of Result

Based on the evaluation results:

1.  **Discriminative Fine-tuning (Classification Head)**:
    *   This approach consistently yields high performance.
    *   **Qwen2:7b** is the top performer with **~85.1% accuracy**.
    *   **Llama3:8b** (~83.3%) and **Mistral:7b** (~80.8%) also perform very well.
    *   Latency is consistently low (~0.15s), making it efficient for pure classification tasks.

2.  **Generative Fine-tuning**:
    *   Performance is more volatile compared to the classification head approach.
    *   **Mistral:7b** adapts best to the generative format for this task, achieving **81.4% accuracy**, which is slightly better than its discriminative counterpart.
    *   **Llama3** and **Qwen2** see a drop in performance (77.6% and 75.3% respectively) compared to their classification head versions.
    *   Several models (Gemma, and unsloth quantized versions of Phi-3/Mistral) struggled significantly, showing near 0% accuracy. This suggests sensitivity to hyperparameters, quantization, or output formatting in the generative setting.

3.  **Overall Conclusion**:
    *   For this specific MATH dataset classification task, **Discriminative Fine-tuning** (Class Head) appears to be the more robust and higher-performing strategy for most models, offering better accuracy and lower latency.
